src/sas/sascalc/dataloader/loader.py

```diff
@@ -61,31 +61,32 @@ def load(self, path, format=None):
         :param format: explicit extension, to force the use
             of a particular reader
 
-        Defaults to the ascii (multi-column) reader
-        if no reader was registered for the file's
-        extension.
+        Defaults to the ascii (multi-column), cansas XML, and cansas NeXuS
+        readers if no reader was registered for the file's extension.
         """
         try:
             return super(Registry, self).load(path, format=format)
-        except:
-            try:
-                # No reader was found. Default to the ascii reader.
-                ascii_loader = ascii_reader.Reader()
-                return ascii_loader.read(path)
-            except:
-                try:
-                    cansas_loader = cansas_reader.Reader()
-                    return cansas_loader.read(path)
-                except:
-                    try:
-                        cansas_nexus_loader = cansas_reader_HDF5.Reader()
-                        return cansas_nexus_loader.read(path)
-                    except:
-                        msg = "\n\tUnknown data format: %s.\n" % path
-                        msg += "\tThe file is not a known format for SasView. "
-                        msg += "The most common formats are multi-column "
-                        msg += "ASCII, CanSAS XML, andCanSAS NeXuS."
-                        raise Exception(msg)
+        except Exception:
+            pass # try the ASCII reader
+        try:
+            ascii_loader = ascii_reader.Reader()
+            return ascii_loader.read(path)
+        except Exception:
+            pass # try the cansas XML reader
+        try:
+            cansas_loader = cansas_reader.Reader()
+            return cansas_loader.read(path)
+        except Exception:
+            pass # try the cansas NeXuS reader
+        try:
+            cansas_nexus_loader = cansas_reader_HDF5.Reader()
+            return cansas_nexus_loader.read(path)
+        except Exception:
+            # No known reader available. Give up and throw an error
+            msg = "\n\tUnknown data format: %s.\n\tThe file is not a " % path
+            msg += "known format for SasView. The most common formats are "
+            msg += "multi-column ASCII, CanSAS XML, and CanSAS NeXuS."
+            raise Exception(msg)
```

Comment:
DataInfo has an 'errors' data member that is supposed to be used to store errors caught while using the default reader for a file being loaded. Unfortunately, the only 3 readers that do this are the cansas XML, cansas HDF5, and saxs ess readers. In loader.py the extra loading attempts are only supposed to happen if the extension registry cannot find a loader for the selected file.

I think making a handful of data loader specific exceptions is the way to go, but I will also have to apply them to all of the readers as well as the loader and extension registry. From there, I should be able to better infer what to do in loader.load() 

Exceptions I think would be useful to add:
NoKnownLoaderException(Exception) - If the extension registry cannot find a reader, throw this to try the 3 default types.
FileContentsException(Exception) - For data files where the registered loader is unable to open or recognize the format, or for failed attempts using the 3 default readers.
DataReaderException(Exception) - For data files that at least partially load, but had issues along the way. These should remain contained within the loader and return them with the data object in DataInfo.errors.

loader.load() would become something like:

    try:
        data = super(Registry, self).load(path, format=format)
        # Exceptions from super.load would skip this return, but data would be set
        return data
    except NoKnownLoaderException as e:
        # Nested try-except to keep e intact
        # The 
        try:
            ascii_loader = ascii_reader.Reader()
            return ascii_loader.read(path)
        except FileContentsException:
            pass # try the cansas XML reader
        # Repeat the above 5 lines for cansas XML and cansas HDF5 reader

    ...

        # If we get here, that means unreadable by SasView and then we can throw this exception
        msg = "\n\tUnknown data format: %s.\n\tThe file is not a " % path
        msg += "known format for SasView. The most common formats are "
        msg += "multi-column ASCII, CanSAS XML, and CanSAS NeXuS."
        raise Exception(msg)
    except FileContentsException as e:
        raise RuntimeException(e)
    except Exception as e:
        # For unknown or unexpected errors
        raise RunTimeException("Some useful message.")
    
    def handle_reader_errors(self, data, e=''):
        """
        Handle any errors thrown by the data loader and try to use as 
        much of the data as possible
        """
        if not isinstance(data, list):
            data = []
        first_data = data[0]
        if not isinstance(first_data, Data1D) and not isinstance(first_data, Data2D):
            x = numpy.zeroes(0)
            y = numpy.zeroes(0)
            first_data = Data1D(x,y)
        first_data.errors.append(str(e))
        data[0] = first_data
        return data

I am closing this pull request while I work on this. What I thought might be a 3 line change is turning into a much larger can of worms.